File: vue.jsonforms/src/entities/owners-and-controllers/CorporateController.ts
import { Director } from "./Director";
import { CompanyDetailsBase } from "../firm-details/CompanyDetailsBase";
import { IndividualController } from "@/entities/owners-and-controllers/IndividualController";
import { ContactNumber } from "@/entities/ContactNumber";
import { IShareholder } from "@/entities/firm-details/IShareholder";

export class CorporateController
  extends CompanyDetailsBase
  implements IShareholder
{
  public firmType: string = "";
  public financialSolvency: string = "";
  public registeredAddress?: string = "";
  public isRegisteredAddressChanged: boolean = false;
  public isTradingAddressSameAsRegisteredAddress: boolean = false;
  public tradingAddress: string = "";
  public isTradingAddressChanged: boolean = false;
  public isHeadOfficeAddressSameAsTradingAddress: boolean = false;
  public headOfficeAddress: string = "";
  public emailAddress: string = "";
  public contactNumber: ContactNumber = new ContactNumber();
  public percentageOfCapital: number | null = null;
  public percentageOfVotingRights: number | null = null;
  public individualOwners: number = 0;
  public corporateOwners: number = 0;
  public isSubjectToRegulationByAnotherRegulator: boolean | undefined;
  public isThirdCountryFirm: boolean | undefined;
  public thirdCountryFirmInfo: string = "";
  public isMemberOfFinancialConglomerate: boolean | undefined;
  public memberOfFinancialConglomerateInfo: string = "";
  public isMemberOfThirdCountryFinancialConglomerate: boolean | undefined;
  public memberOfThirdCountryFinancialConglomerateInfo: string = "";
  public isMemberOfThirdCountryBanking: boolean | undefined;
  public memberOfThirdCountryBankingInfo: string = "";
  public hasBeenSubjectToAnyMaterialComplaints: boolean | undefined;
  public beenSubjectToAnyMaterialComplaintsInfo: string = "";
  public directors: Director[] = [] as Director[];
  public individualControllers: IndividualController[] =
    [] as IndividualController[];
  public corporateControllers: CorporateController[] =
    [] as CorporateController[];
  public supportingDocumentsUrls: string[] = [];
}

File: CSharp.Functions/Common/Entities/FirmDetails/OwnersAndControllers/CorporateController.cs
namespace Common.Entities
{
    public class CorporateController : CompanyDetails
    {
        public string? FirmType { get; set; }
        public string? FinancialSolvency { get; set; }
        public string? RegisteredAddress { get; set; }
        public bool IsRegisteredAddressChanged { get; set; }
        public bool IsTradingAddressSameAsRegisteredAddress { get; set; }
        public string? TradingAddress { get; set; }
        public bool IsTradingAddressChanged { get; set; }
        public bool IsHeadOfficeAddressSameAsTradingAddress { get; set; }
        public string? HeadOfficeAddress { get; set; }
        public string? EmailAddress { get; set; }
        public ContactNumber? ContactNumber { get; set; }
        public double? PercentageOfCapital { get; set; }
        public double? PercentageOfVotingRights { get; set; }
        public double IndividualOwners { get; set; }
        public int CorporateOwners { get; set; }
        public bool IsSubjectToRegulationByAnotherRegulator { get; set; }
        public bool IsThirdCountryFirm { get; set; }
        public string? ThirdCountryFirmInfo { get; set; }
        public bool IsMemberOfFinancialConglomerate { get; set; }
        public string? MemberOfFinancialConglomerateInfo { get; set; }
        public bool IsMemberOfThirdCountryFinancialConglomerate { get; set; }
        public string? MemberOfThirdCountryFinancialConglomerateInfo { get; set; }
        public bool IsMemberOfThirdCountryBanking { get; set; }
        public string? MemberOfThirdCountryBankingInfo { get; set; }
        public bool HasBeenSubjectToAnyMaterialComplaints { get; set; }
        public string? BeenSubjectToAnyMaterialComplaintsInfo { get; set; }
        public List<Director?> Directors { get; set; } = new();
        public List<IndividualController> IndividualControllers { get; set; } = new();
        public List<CorporateController> CorporateControllers { get; set; } = new();
        public List<string> SupportingDocumentsUrls { get; set; } = new();
    }
}